File: web/src/lib/features.ts
/**
 * GridGuard TypeScript feature engine.
 *
 * MIRRORS src/features.py (Python reference) EXACTLY — same imputation, same
 * formulas, same guards, same feature order. Cross-language parity is asserted
 * by tests/data/parity_cases.json goldens produced by scripts/prepare.py.
 * Any change here must be mirrored in Python and regenerate the goldens.
 */

export const FEATURE_SPEC_VERSION = 2;
export const DEFAULT_DOW0 = 3; // JS getDay of 2014-01-01 (Wed)
const MONTH_BUCKET = 30;
const EPS = 1e-9;

export type Maybe = number | null;

export const FEATURE_NAMES: string[] = [
  "mean_cons", "std_cons", "min_cons", "q25_cons", "median_cons",
  "q75_cons", "max_cons", "cv_cons",
  "skew_cons", "kurt_cons",
  "zero_ratio", "long_zero_run",
  "weekday_mean", "weekend_mean", "weekend_ratio",
  "month_spread",
  "half_ratio", "trend_slope",
  "mad_daily", "std_diff", "pct_change_mean",
  "zero_run_norm",
];

export function impute(series: Maybe[]): number[] {
  const xs: (number | null)[] = series.map((v) =>
    v === null ? null : v < 0 ? 0 : v
  );
  const n = xs.length;
  if (n === 0) return [];
  let fv = -1;
  for (let i = 0; i < n; i++) if (xs[i] !== null) { fv = i; break; }
  if (fv === -1) return new Array<number>(n).fill(0);
  for (let k = 0; k < fv; k++) xs[k] = xs[fv];
  let i = fv;
  while (i < n) {
    if (xs[i] === null) {
      let j = i;
      while (j < n && xs[j] === null) j++;
      const left = xs[i - 1] as number;
      if (j === n) {
        for (let k = i; k < j; k++) xs[k] = left;
      } else {
        const right = xs[j] as number;
        const gap = j - i;
        for (let k = i; k < j; k++)
          xs[k] = left + ((right - left) * (k - i + 1)) / (gap + 1);
      }
      i = j;
    } else {
      i++;
    }
  }
  return xs as number[];
}

const mean = (xs: number[]): number =>
  xs.length ? xs.reduce((a, b) => a + b, 0) / xs.length : 0;

function std(xs: number[]): number {
  if (xs.length < 2) return 0;
  const m = mean(xs);
  let s = 0;
  for (const x of xs) s += (x - m) * (x - m);
  return Math.sqrt(s / (xs.length - 1));
}

function skew(xs: number[]): number {
  const n = xs.length;
  if (n < 3) return 0;
  const m = mean(xs);
  let m2 = 0, m3 = 0;
  for (const x of xs) {
    const d = x - m;
    m2 += d * d;
    m3 += d * d * d;
  }
  m2 /= n; m3 /= n;
  return m2 > 1e-24 ? m3 / Math.pow(m2, 1.5) : 0;
}

function kurt(xs: number[]): number {
  const n = xs.length;
  if (n < 4) return 0;
  const m = mean(xs);
  let m2 = 0, m4 = 0;
  for (const x of xs) {
    const d = x - m;
    m2 += d * d;
    m4 += d * d * d * d;
  }
  m2 /= n; m4 /= n;
  return m2 > 1e-24 ? m4 / (m2 * m2) - 3 : 0;
}

function quantile(xs: number[], p: number): number {
  if (xs.length === 0) return 0;
  const ys = [...xs].sort((a, b) => a - b);
  if (ys.length === 1) return ys[0];
  const h = (ys.length - 1) * p;
  const lo = Math.floor(h);
  const hi = Math.min(lo + 1, ys.length - 1);
  const frac = h - lo;
  return ys[lo] * (1 - frac) + ys[hi] * frac;
}

const cov = (xs: number[]): number => {
  const m = mean(xs);
  return m > 1e-12 ? std(xs) / m : 0;
};

const zeroRatio = (xs: number[]): number =>
  xs.length ? xs.filter((x) => x <= EPS).length / xs.length : 0;

function longestZeroRun(xs: number[]): number {
  let best = 0, run = 0;
  for (const x of xs) {
    if (x <= EPS) {
      run++;
      if (run > best) best = run;
    } else run = 0;
  }
  return best;
}

function slope(xs: number[]): number {
  const n = xs.length;
  if (n < 2) return 0;
  const mxi = (n - 1) / 2;
  const my = mean(xs);
  let num = 0;
  for (let i = 0; i < n; i++) num += (i - mxi) * (xs[i] - my);
  let den = 0;
  for (let i = 0; i < n; i++) den += (i - mxi) * (i - mxi);
  return den > 1e-12 ? num / den : 0;
}

function meanAbsDiff(a: number[], b: number[]): number {
  if (!a.length) return 0;
  let s = 0;
  for (let i = 0; i < a.length; i++) s += Math.abs(b[i] - a[i]);
  return s / a.length;
}

function meanAbsPctChange(a: number[], b: number[]): number {
  if (!a.length) return 0;
  const vals: number[] = [];
  for (let i = 0; i < a.length; i++)
    if (Math.abs(a[i]) > 1e-9) vals.push(Math.abs(b[i] - a[i]) / Math.abs(a[i]));
  return vals.length ? mean(vals) : 0;
}

export function featureVector(series: Maybe[], dow0: number = DEFAULT_DOW0): number[] {
  const x = impute(series);
  const n = x.length;
  if (n === 0) return FEATURE_NAMES.map(() => 0);

  const mu = mean(x);
  const sd = std(x);

  const wd: number[] = [], we: number[] = [];
  for (let i = 0; i < n; i++) {
    const dow = (dow0 + i) % 7;
    (dow === 0 || dow === 6 ? we : wd).push(x[i]);
  }
  const wdMu = wd.length ? mean(wd) : 0;
  const weMu = we.length ? mean(we) : 0;
  const weRatio = wdMu > 1e-12 ? weMu / wdMu : weMu <= 1e-12 ? 1 : 2;

  const nb = Math.ceil(n / MONTH_BUCKET);
  const bucketMeans: number[] = [];
  for (let b = 0; b < nb; b++) {
    const seg = x.slice(b * MONTH_BUCKET, (b + 1) * MONTH_BUCKET);
    if (seg.length) bucketMeans.push(mean(seg));
  }
  const monthSpread =
    bucketMeans.length >= 2 && mu > 1e-12
      ? (Math.max(...bucketMeans) - Math.min(...bucketMeans)) / mu
      : 0;

  const h = Math.floor(n / 2);
  const m1 = mean(x.slice(0, h));
  const m2 = h > 0 ? mean(x.slice(h)) : 0;
  const halfRatio = m1 > 1e-12 ? m2 / m1 : m2 <= 1e-12 ? 1 : 2;

  let prev: number[], nxt: number[], diffs: number[];
  if (n > 1) {
    prev = x.slice(0, n - 1);
    nxt = x.slice(1);
    diffs = nxt.map((v, i) => v - prev[i]);
  } else {
    prev = nxt = diffs = [0];
  }

  const lzr = longestZeroRun(x);

  return [
    mu, sd, Math.min(...x), quantile(x, 0.25), quantile(x, 0.5),
    quantile(x, 0.75), Math.max(...x),
    cov(x),
    skew(x), kurt(x),
    zeroRatio(x), lzr,
    wdMu, weMu, weRatio,
    monthSpread,
    halfRatio, slope(x),
    meanAbsDiff(prev, nxt), std(diffs), meanAbsPctChange(prev, nxt),
    lzr / n,
  ];
}

File: src/features.py
"""Declarative feature engineering for smart-meter consumption series.

The SAME definition drives (a) Python training/prediction and (b) the
TypeScript web app: `export_spec()` -> JSON -> web/src/lib/features.ts.
One source of truth guarantees cross-language parity; tests assert it
bit-for-bit on random series.

Each consumer is a daily consumption series x[0..n-1] (kWh) with possible
missing days (None/NaN). Imputation happens BEFORE any statistic.

22 features, families:
  A. Consumption level   mean, std, min, q25, median, q75, max, CV
  B. Shape / tails       skew, kurtosis (population moments)
  C. Zero behaviour      zero_ratio, longest zero run (+ normalized)
  D. Temporal            weekday/weekend means, weekend ratio, month spread
  E. Drift               first-half vs second-half ratio, OLS trend slope
  F. Volatility          mean abs daily diff, std of diffs, mean abs pct change
"""
from __future__ import annotations

import json
import math
from typing import Any, Dict, List, Optional

FEATURE_SPEC_VERSION = 2

# Weekday index (JS Date.getDay convention, 0=Sunday) of day 0 of the series.
# SGCC series starts 2014-01-01 = Wednesday = 3.
DEFAULT_DOW0 = 3

MONTH_BUCKET = 30  # days per "month" bucket for month_spread

# ---------------------------------------------------------------------------
# Spec (single source of truth consumed by the TypeScript implementation)
# ---------------------------------------------------------------------------

def build_imputer_spec() -> Dict[str, Any]:
    return {
        "kind": "linear_interp_clip0",
        "interior": "linear",
        "edge": "nearest",
        "all_missing": 0.0,
    }


def export_spec(dow0: int = DEFAULT_DOW0) -> Dict[str, Any]:
    return {
        "version": FEATURE_SPEC_VERSION,
        "imputer": build_imputer_spec(),
        "dow0": dow0,
        "month_bucket": MONTH_BUCKET,
        "features": feature_names(),
    }


def spec_json(dow0: int = DEFAULT_DOW0) -> str:
    return json.dumps(export_spec(dow0=dow0), separators=(",", ":"))


# ---------------------------------------------------------------------------
# Imputation
# ---------------------------------------------------------------------------

def impute(series: List[Optional[float]]) -> List[float]:
    """clip negatives to 0 -> linear interpolation over interior gaps ->
    nearest-neighbour fill for edge gaps -> 0 for an all-missing series."""
    xs: List[Optional[float]] = [
        (0.0 if v < 0 else float(v)) if v is not None else None for v in series
    ]
    n = len(xs)
    if n == 0:
        return []
    fv = next((i for i, v in enumerate(xs) if v is not None), None)
    if fv is None:
        return [0.0] * n
    for k in range(fv):                       # leading gap <- first valid
        xs[k] = xs[fv]
    i = fv
    while i < n:
        if xs[i] is None:
            j = i
            while j < n and xs[j] is None:
                j += 1
            left = xs[i - 1]
            if j == n:                        # trailing gap -> last valid
                for k in range(i, j):
                    xs[k] = left
            else:                             # interior gap -> linear interp
                right = xs[j]
                gap = j - i
                for k in range(i, j):
                    xs[k] = left + (right - left) * (k - i + 1) / (gap + 1)
            i = j
        else:
            i += 1
    return [float(v) for v in xs]


# ---------------------------------------------------------------------------
# Statistical primitives (pure; TS mirrors exactly)
# ---------------------------------------------------------------------------

def _mean(xs: List[float]) -> float:
    return sum(xs) / len(xs) if xs else 0.0


def _std(xs: List[float]) -> float:
    if len(xs) < 2:
        return 0.0
    m = _mean(xs)
    return math.sqrt(sum((x - m) ** 2 for x in xs) / (len(xs) - 1))


def _skew(xs: List[float]) -> float:
    """Population skewness g1 = m3 / m2^1.5."""
    n = len(xs)
    if n < 3:
        return 0.0
    m = _mean(xs)
    m2 = sum((x - m) ** 2 for x in xs) / n
    m3 = sum((x - m) ** 3 for x in xs) / n
    return m3 / (m2 ** 1.5) if m2 > 1e-24 else 0.0


def _kurt(xs: List[float]) -> float:
    """Population excess kurtosis g2 = m4/m2^2 - 3."""
    n = len(xs)
    if n < 4:
        return 0.0
    m = _mean(xs)
    m2 = sum((x - m) ** 2 for x in xs) / n
    m4 = sum((x - m) ** 4 for x in xs) / n
    return m4 / (m2 ** 2) - 3.0 if m2 > 1e-24 else 0.0


def _q(xs: List[float], p: float) -> float:
    """Linear-interpolated quantile (numpy default)."""
    if not xs:
        return 0.0
    ys = sorted(xs)
    if len(ys) == 1:
        return ys[0]
    h = (len(ys) - 1) * p
    lo = int(math.floor(h))
    hi = min(lo + 1, len(ys) - 1)
    frac = h - lo
    return ys[lo] * (1 - frac) + ys[hi] * frac


def _cov(xs: List[float]) -> float:
    m = _mean(xs)
    return (_std(xs) / m) if m > 1e-12 else 0.0


def _zero_ratio(xs: List[float]) -> float:
    if not xs:
        return 0.0
    return sum(1 for x in xs if x <= 1e-9) / len(xs)


def _longest_zero_run(xs: List[float]) -> int:
    best = run = 0
    for x in xs:
        if x <= 1e-9:
            run += 1
            if run > best:
                best = run
        else:
            run = 0
    return best


def _slope(xs: List[float]) -> float:
    """OLS slope vs day index."""
    n = len(xs)
    if n < 2:
        return 0.0
    mx = (n - 1) / 2.0
    my = _mean(xs)
    num = sum((i - mx) * (y - my) for i, y in enumerate(xs))
    den = sum((i - mx) ** 2 for i in range(n))
    return num / den if den > 1e-12 else 0.0


def _mean_abs_diff(a: List[float], b: List[float]) -> float:
    if not a:
        return 0.0
    return sum(abs(u - v) for u, v in zip(a, b)) / len(a)


def _mean_abs_pct_change(a: List[float], b: List[float]) -> float:
    if not a:
        return 0.0
    vals = [abs(v - u) / abs(u) for u, v in zip(a, b) if abs(u) > 1e-9]
    return _mean(vals) if vals else 0.0


# ---------------------------------------------------------------------------
# Feature computation
# ---------------------------------------------------------------------------

def feature_names() -> List[str]:
    return [
        "mean_cons", "std_cons", "min_cons", "q25_cons", "median_cons",
        "q75_cons", "max_cons", "cv_cons",
        "skew_cons", "kurt_cons",
        "zero_ratio", "long_zero_run",
        "weekday_mean", "weekend_mean", "weekend_ratio",
        "month_spread",
        "half_ratio", "trend_slope",
        "mad_daily", "std_diff", "pct_change_mean",
        "zero_run_norm",
    ]


def feature_vector(series: List[Optional[float]], dow0: int = DEFAULT_DOW0) -> List[float]:
    x = impute(series)
    n = len(x)
    if n == 0:
        return [0.0] * len(feature_names())

    mu = _mean(x)
    sd = _std(x)

    wd = [x[i] for i in range(n) if ((dow0 + i) % 7) not in (0, 6)]
    we = [x[i] for i in range(n) if ((dow0 + i) % 7) in (0, 6)]
    wd_mu = _mean(wd)
    we_mu = _mean(we)
    if wd_mu > 1e-12:
        we_ratio = we_mu / wd_mu
    else:
        we_ratio = 1.0 if we_mu <= 1e-12 else 2.0

    bs = MONTH_BUCKET
    buckets = [x[i * bs:(i + 1) * bs] for i in range((n + bs - 1) // bs)]
    bucket_means = [_mean(b) for b in buckets if b]
    if len(bucket_means) >= 2 and mu > 1e-12:
        month_spread = (max(bucket_means) - min(bucket_means)) / mu
    else:
        month_spread = 0.0

    h = n // 2
    m1 = _mean(x[:h])
    m2 = _mean(x[h:]) if h > 0 else 0.0
    if m1 > 1e-12:
        half_ratio = m2 / m1
    else:
        half_ratio = 1.0 if m2 <= 1e-12 else 2.0

    if n > 1:
        prev = x[:-1]
        nxt = x[1:]
        diffs = [b - a for a, b in zip(prev, nxt)]
    else:
        prev = nxt = diffs = [0.0]

    lzr = _longest_zero_run(x)

    return [
        mu, sd, min(x), _q(x, 0.25), _q(x, 0.5), _q(x, 0.75), max(x),
        _cov(x),
        _skew(x), _kurt(x),
        _zero_ratio(x), float(lzr),
        wd_mu, we_mu, we_ratio,
        month_spread,
        half_ratio, _slope(x),
        _mean_abs_diff(prev, nxt), _std(diffs), _mean_abs_pct_change(prev, nxt),
        lzr / n,
    ]


def features_from_rows(rows: List[List[Optional[float]]], dow0: int = DEFAULT_DOW0) -> List[List[float]]:
    return [feature_vector(r, dow0=dow0) for r in rows]
